crates/stages/src/stages/hashing_storage.rs

```diff
@@ -0,0 +1,491 @@
+use tracing::*;
+
+const STORAGE_HASHING: StageId = StageId("StorageHashingStage");
+
+/// Storage hashing stage hashes plain storage.
+/// This is preparation before generating intermediate hashes and calculating Merkle tree root.
+#[derive(Debug)]
+pub struct StorageHashingStage {
+    /// The threshold for switching from incremental hashing
+    /// of changes to whole storage hashing. Num of transitions.
+    pub clean_threshold: u64,
+    /// The size of inserted items after which the control
+    /// flow will be returned to the pipeline for commit
+    pub commit_threshold: u64,
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for StorageHashingStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        STORAGE_HASHING
+    }
+
+    /// Execute the stage.
+    async fn execute(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let previous_stage_progress = input.previous_stage_progress();
+
+        // read storage changeset, merge it into one changeset and calculate storage hashes.
+        let from_transition = tx.get_block_transition(stage_progress)? + 1;
+        let to_transition = tx.get_block_transition(previous_stage_progress)? + 1;
+
+        // if there are more blocks then threshold it is faster to go over Plain state and hash all
+        // account otherwise take changesets aggregate the sets and apply hashing to
+        // AccountHashing table
+        if to_transition - from_transition > self.clean_threshold {
+            tx.clear::<tables::HashedStorage>()?;
+            tx.commit()?;
+
+            let mut first_key = H160::zero();
+            loop {
+                let next_key = {
+                    let mut storage = tx.cursor_dup_read::<tables::PlainStorageState>()?;
+
+                    let hashed_batch = storage
+                        .walk(first_key)?
+                        .take(self.commit_threshold as usize)
+                        .map(|res| {
+                            res.map(|(address, mut slot)| {
+                                // both account address and storage slot key are hashed for merkle
+                                // tree.
+                                slot.key = keccak256(slot.key);
+                                (keccak256(address), slot)
+                            })
+                        })
+                        .collect::<Result<BTreeMap<_, _>, _>>()?;
+
+                    // next key of iterator
+                    let next_key = storage.next()?;
+
+                    let mut hashes = tx.cursor_write::<tables::HashedStorage>()?;
+                    // iterate and append presorted hashed slots
+                    hashed_batch.into_iter().try_for_each(|(k, v)| hashes.append(k, v))?;
+                    next_key
+                };
+                tx.commit()?;
```

Comment:
Do we want to be committing inside the stage? Thought we want to avoid commits internally and return control flow to the pipeline once we hit commit threshold?

That would also mean we could remove the loop above?

Reply:
Yes, if we return, there is no way to save progress. Here, we are iterating over all keys in the table, not over the blocks as in other stages.